### DISCUSSION TITLE
Number of iterations/hops in NeighorLoader

### DISCUSSION THREAD

**Discussion by MDrance** [pyg-team/pytorch_geometric#5488] [Q&A] [answered]
Hello,

I'm a little bit confused about how to use the number of iterations/hops when training my model. In my understanding, the number of iterations/hops is by default dependent of the number of layers in my model, so if I used and encoder with two _**GraphConv**_ layers, the information used to update the embedding of node _**u**_ will come from its direct neighbors and their neighbors too.
In the _**NeighborLoader**_ documentation I found this :

```
from torch_geometric.datasets import Planetoid
from torch_geometric.loader import NeighborLoader

data = Planetoid(path, name='Cora')[0]

loader = NeighborLoader(
    data,
    # Sample 30 neighbors for each node for 2 iterations
    num_neighbors=[30] * 2,
    # Use a batch size of 128 for sampling training nodes
    batch_size=128,
    input_nodes=data.train_mask,
)

sampled_data = next(iter(loader))
print(sampled_data.batch_size)
128
```
Where it states that we have to specify _**num\_neighbors=[30] * 2,**_ to use two iterations\hops from the current node.
Am I misunderstanding something here ?

Best,
Martin

**Comment by rusty1s** [comment, marked as answer]
This is correct. `num_neighbors=[30, 30]` will indicate that you will sample for two-hops while sampling 30 neighbors per node each. This should mimic the number of GNN layers.

**Reply by MDrance** [reply]
Hello Matthias thank you for your answer.
It's clear now for me about the `NeighborLoader`. But what happen if I only pass `num_neighbors=[30]` in a model with two layers ? Or conversely, I pass `num_neighbors=[30, 30]` in a model with only one layer ?

**Reply by rusty1s** [reply]
Yes, this will usually lead to wrong computation since nodes sampled in the last hop will not receive any information during message passing, and thus sending empty information to nodes in the second layer. We provide an `directed=False` option in the `NeighborLoader` to circumvent this. That is, we create an induced subgraph containing all samples nodes, such that message passing with a larger number of layers is still well-defined.

**Reply by MDrance** [reply]
Got it thanks for the explanation !

**Reply by MDrance** [reply]
Since I opened a discussion I would like to ask an other question, I'm using a `GAE` model to reconstruct nodes embeddings I already have:
- In a transductive setting, is `RandomNodeSplit` the best tool to use to split my `Data` object ? Should I use instead `RandomLinkSplit` as I don't mind having the same nodes in my train/val/test split ? Or should I use training data during training and then make the test on all the graph as it is the case on the `autoencoder.py` example ?

**Reply by rusty1s** [reply]
For transductive link prediction tasks, you would want to use `RandomLinkSplit` rather than `RandomNodeSplit` such that you split by edges rather than by nodes.

**Reply by MDrance** [reply]
So `RandomLinkSplit` is to be used for edges based tasks, `RandoNodeSplit` for nodes based task as you use `NeighborLoader` for node based task and `LinkNeighborLoader` for edges based task.

I think I get it, thanks a lot for your answers,
Martin
